20231214-listUsers-api.md

```diff
@@ -0,0 +1,887 @@
+| group:eng      | member   | group:fga#member      |
+| group:fga      | member   | user:andres           |
+| group:fga      | member   | group:fga-core#member |
+| group:fga-core | member   | user:jon              |
+
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filter: [{type: "user"}]
+}) --> ["user:jon", "user:andres"]
+```
+This example requires recursive expansion.
+> ℹ️ The usage of `expand` from here forward refers to an internal expand function, and it should not be confused with the public Expand API, but it operates somewhat similarly.
+
+
+1. Expand the set of subjects/users that relate to `document:1#viewer`. 
+
+   `expand(document:1#viewer)` --> ["group:eng#member"]
+
+1. Expand the new set of subjects/users that make up `group:eng#member` set.
+
+   `expand(group:eng#member)` --> ["group:fga#member"]
+
+1. Expand the set of subjects/users that make up the `group:fga#member` set.
+
+   `expand(group:fga#member)` --> ["user:andres", "group:fga-core#member"]
+
+   1. We find a terminal/concrete object of `user:andres`, which matches the `user` in the user filter, and so we add it to the list of items to include in the response.
+
+1. Continue expanding the residual set of subjects/users in `group:fga-core#member`
+
+   `expand(group:fga-core#member)` --> ["user:jon"]
+
+   1. We find a terminal/concrete object of `user:jon`, which matches the `user` in the user filter, so add it to the list of items to include in the response
+
+1. No further subjects to expand, and so we're done. Return the items we accumulated from the steps above.
+
+Visually, the overall recursive call tree looks like the following:
+
+```
+expand(document:1#viewer) --> ["group:eng#member"]
+|-> expand(group:eng#member) --> ["group:fga#member"]
+|---> expand(group:fga#member) --> ["user:andres", "group:fga-core#member"]
+      yield "user:andres"
+|-----> expand(group:fga-core#member) --> ["user:jon"]
+        yield "user:jon"
+
+return ["user:andres", "user:jon"]
+```
+
+### Example 1c (Direct Relation with Typed Public Wildcard)
+See [Typed Public Wildcard](#typed-public-wildcard) section for more context.
+
+```
+model
+  schema 1.1
+
+type user
+
+type document
+  relations
+    define viewer: [user:*]
+```
+| object     | relation | user   |
+|------------|----------|--------|
+| document:1 | viewer   | user:* |
+
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filter: [
+    {
+      type: "user"
+    }
+  ]
+}) --> {
+  users: ["user"],
+  wildcard_types: []
+}
+```
+
+1. Expand the set of subjects/users that relate to `document:1#viewer`.
+
+   `expand(document:1#viewer)` --> ["user:*"]
+
+    1. We find a terminal/concrete wildcard of `user:*`, which matches the `user:*` in the user filter, and so we add it to the list of items to include in the response.
+
+1. No further subjects to expand, and so we're done. Return the items we accumulated from the steps above.
```

Comment:
```suggestion
1. No further users to expand, and so we're done. Return the items we accumulated from the steps above.
```

We use the term user everywhere, there is no need to suddenly switch to subject

Reply:
I used the term subject specifically because the type is overlapping here. The example is distinguishing between `user`, `user:*`, and the "user" objects being expanded. I wanted to avoid overloading the term `user` here.